ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py

```diff
@@ -31,49 +33,103 @@ def run(argv):
+        calls_parquet = f"{out_dir}/calls.snps.pass.parquet"
+        featuremap_on_calls_vcf = f"{out_dir}/featuremap_on_calls.vcf.gz"
+        featuremap_on_calls_parquet = f"{out_dir}/featuremap_on_calls.parquet"
+        qual_af_clusters_on_calls = f"{out_dir}/qual_af_clusters_on_calls.parquet"
         sp.print_and_run(f"bcftools view --type snps -f PASS -m2 -M2 {unfiltered_calls} -Oz -o {calls_vcf}")
         sp.print_and_run(
             f"bedtools intersect -a {featuremap_vcf} -b {calls_vcf} -header | "
-            f"bcftools view -Oz -o {featuremap_on_calls}"
+            f"bcftools view -Oz -o {featuremap_on_calls_vcf}"
         )
-        featuremap_vcf = featuremap_on_calls
-        sp.print_and_run(f"bcftools query -f '%CHROM\t%POS0\t%POS\t[%VAF]\\n' {calls_vcf} | grep -v , > {afs_bed}")
-    else:
-        sp.print_and_run(
-            "bcftools query -f '%CHROM\t%POS0\t%POS\t%X_READ_COUNT' "
-            " | bedtools groupby -c 3 -full -o count | awk -v OFS='\t' '{print $1,$2,$3,$5/$4}' " + f" > {afs_bed}"
+
+        sp.print_and_run_clt(
+            featuremap_to_dataframe.run,
+            [],
+            {
+                "-i": calls_vcf,
+                "-r": args.ref_fasta,
+                "-f": "TGCA",
+                "--info_fields_override": "",
+                "--format-fields": "DP=int AD=str SB=str VAF=str PL=str BG_AD=str",
+                "-o": calls_parquet,
+            },
+            flags=["--report-bases-in-reference-direction"],
         )
 
-    for bed_file, vcf_file, parquet_file, min_af, max_af in (tp_args, fp_args):
-        sp.print_and_run(f"awk '$4 >= {min_af} && $4 <= {max_af}' {afs_bed} > {bed_file}")
         sp.print_and_run(
-            f"bedtools intersect -a {featuremap_vcf} -b {bed_file} -u -header | " f"bcftools view -Oz -o {vcf_file}"
+            f"papermill {nb_dir}/qual_af_clustering.ipynb "
+            f"{out_dir}/qual_af_clustering.nbconvert.ipynb "
+            f"-p vcf_input {calls_parquet} "
+            f"-p output_file {qual_af_clusters_on_calls} "
+        )
+        sp.print_and_run(
+            "jupyter nbconvert --to html "
+            f"{out_dir}/qual_af_clustering.nbconvert.ipynb "
+            f"--no-input --output qual_af_clustering.html"
         )
         sp.print_and_run_clt(
-            featuremap_to_dataframe.run, [], {"-i": vcf_file, "-r": args.ref_fasta, "-f": "TGCA", "-o": parquet_file}
+            featuremap_to_dataframe.run,
+            [],
+            {"-i": featuremap_on_calls_vcf, "-r": args.ref_fasta, "-f": "TGCA", "-o": featuremap_on_calls_parquet},
+        )
+        qual_af_df = pd.read_parquet(qual_af_clusters_on_calls)
+        featuremap_on_calls_df = pd.read_parquet(featuremap_on_calls_parquet)
+        featuremap_on_calls_df = featuremap_on_calls_df.drop(columns=["qual", "filter"])
+        qual_af_df.merge(featuremap_on_calls_df, on=["chrom", "pos", "ref", "alt"]).to_parquet(labeled_training_set)
```

Comment:
Does that work in a reasonable time? I would worry about a merge on 4 columns. Setting as index and running "join" is much faster than "merge" on columns in my experience, and I think you can safely drop the "ref". Maybe even the "chrom"?

Reply:
BTW, if you do set an index with multiple columns it's best to reset it before saving

Reply:
Yes, it is quite fast. 
We are only looking at pass-filter loci here.
I haven't noticed any problem with the index of the labeled_training_set, but will look again in the analysis that uses it.

Reply:
The index looks fine to me:
![image](https://github.com/Ultimagen/VariantCalling/assets/44941371/baa07f88-06c9-4171-a0dd-47bf86eef964)

Reply:
I meant that if you use a multiindex like I suggested it's best not to save this way, nvm